Fix submenu justification and spacer orientation.

tellthemachines opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

#36169 broke submenu alignment when the Navigation block is aligned right or space-between. This PR adds back the missing classnames as a temporary fix until we can sort out a better way to align submenus (#36241) 

It also patches the orientation context that Spacer block needs to work correctly inside Navigation. Again, this is a temporary measure until we find a better way to do this (#36197)

In doing this I noticed the Navigation block still has [an `is-vertical` classname](https://github.com/WordPress/gutenberg/blob/fix/submenu-alignment/packages/block-library/src/navigation/edit/index.js#L201) to which a few editor styles are added, mainly to do with the placeholder. I'm thinking these might no longer be needed since vertical/horizontal is now a setting instead of a block variation, so the placeholder will, in principle, never appear in vertical form. @jasmussen could we remove these styles?

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

Add a Navigation block with some nested submenus (at least two levels of nesting). 
Align it right, and verify that 2nd level nested submenus are opening to the left instead of to the right.

Add a Spacer block inside the Navigation. 
Verify that in horizontal orientation, the Spacer grows horizontally.

## Screenshots <!-- if applicable -->

## Types of changes
<!-- What types of changes does your code introduce?  -->
Bug fix (non-breaking change which fixes an issue)
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

tellthemachines committed:
Fix submenu justification and spacer orientation.

jasmussen commented:
Really nice, thank you. This fixes it in the editor right off the bat:
![nav justifications](https://user-images.githubusercontent.com/1204802/140892153-941a0672-9d13-43dc-b4fe-7bac05ac0932.gif)

It doesn't appear to be working on the frontend:

<img width="1726" alt="Screenshot 2021-11-09 at 09 48 18" src="https://user-images.githubusercontent.com/1204802/140892187-c965bc60-a47f-4542-a5d5-ed4227922243.png">

I'll dive in and see if I can push a fix for that.

jasmussen commented:
> In doing this I noticed the Navigation block still has an is-vertical classname to which a few editor styles are added, mainly to do with the placeholder. I'm thinking these might no longer be needed since vertical/horizontal is now a setting instead of a block variation, so the placeholder will, in principle, never appear in vertical form. @jasmussen could we remove these styles?

About the placeholder, that's a good question, and goes to the heart of how we should design placeholders. The thing is, in a site editing context, the placeholder should ideally have a footprint that is _identical_ to that of the final result. Without it, [it will cause layout shifts](https://github.com/WordPress/gutenberg/issues/20853#issuecomment-621030013). 

There's wiggleroom of course — there are some blocks where the pre-configuraiton of key values is so crucial that it deserves up-front _weight_, but even then we have to consider the context in which they will be used, such as small viewports, or of the block is inserted inside a thin column. For example the navigation block in a column is already leveraging the built-in `is-small` classname that is being applied to the `Placeholder` by a resizeobserver:

<img width="770" alt="Screenshot 2021-11-09 at 09 55 03" src="https://user-images.githubusercontent.com/1204802/140893444-9f16ddef-befd-485f-9c9b-6ae91c5e5dab.png">

As far as block placeholders that look like the end result, my current favorite is the Site Logo:
![site logo](https://user-images.githubusercontent.com/1204802/140893634-0b7efa25-e8bf-4a9c-bd6e-3e9734e3c6d1.gif)

Its footprint is identical, and it inherit colors and styles from the theme even in its unconfigured state. 

So back to your question, can we remove the `is-vertical` styles? **Yes, probably,** because most block setup states, _including navigation_, need to be revisited with a fresh perspective based on the challenges outlined above. In that vein, it's okay to pull back on the customization in the mean time. (Tangentially, #36302 is one tiny experiment in further reducing the footprint of the navigation setup state).

jasmussen commented:
> It also patches the orientation context that Spacer block needs to work correctly inside Navigation. Again, this is a temporary measure until we find a better way to do this (#36197)

Thanks for working on this. I'm not sure what to look for, as trunk and this branch appear the same to me:
![spacer](https://user-images.githubusercontent.com/1204802/140894505-a04cd75e-af82-4267-af41-c2b21050ad22.gif)

Some issues:

- Regardless of orientation, I can only resize the height. Likely the orientation related issue you've tracked in #36197.
- It's only 1px wide when freshly inserted, so it's hard to select. Upon clicking it, a width is set so it's easier to select.

It also made me realize that the mover controls are now all vertical, for all blocks inside the container, despite having horizontal orientation. This appears to be a general challenge for flex orientation blocks, as the Buttons block also shows vertical movers when it should be showing horizontal ones:

<img width="668" alt="Screenshot 2021-11-09 at 10 05 47" src="https://user-images.githubusercontent.com/1204802/140894887-43edb7dc-2d58-4244-9f9b-432d0311527c.png">

Do you know if this is tracked separately? In any case, since these are general issues for the spacer and mover controls, they should not block this PR from fixing the justifications.

jasmussen commented:
Back to the justifications which work in the editor but not in the frontend. It appears the classnames are working perfectly, the missing piece is that the `items-justified-n` classes are not output on the frontend. If I edit them in, things work:

![inspectin](https://user-images.githubusercontent.com/1204802/140895482-db4d0e97-e7d2-45bf-9514-53ee5f2f32fb.gif)

I suspect that's beyond my skill to fix. But it seems like once we add those classes, we can land this one. Thanks again!

tellthemachines committed:
Add back justification classes in the front end.

tellthemachines committed:
Add default orientation for Spacer.

tellthemachines committed:
Remove legacy attributes and styles.

tellthemachines commented:
Ok I've fixed the justification classes on the front end, and the Spacer issue (there was no default value for orientation; I added one in so it should always work now). Also removed the `is-vertical` styles.

jasmussen commented:
I'm having some local development environments that make me unable to test latest trunk, or this PR, at the moment. But if this works on the frontend, it'd be good to land, we can always follow up on things. It would be good with a quick sanity check by another.

tellthemachines committed:
Fix php lint errors.

tellthemachines committed:
Mark orientation change as not persistent.

jasmussen reviewed (approved):
Works well:
<img width="900" alt="Screenshot 2021-11-11 at 08 37 14" src="https://user-images.githubusercontent.com/1204802/141257607-4df37814-21bf-4136-99cb-8becc08c3e89.png">
<img width="845" alt="Screenshot 2021-11-11 at 08 37 24" src="https://user-images.githubusercontent.com/1204802/141257612-2c8d0817-3e63-476e-ba8d-c5427d8c28c3.png">

Thank you for the cleanup as well. As noted, I think I might be able to restore some of the vertical behavior for the placeholder when the orientation is changed — whether that's desirable remains to be seen, but no reason to lug the cruft along with it in the mean time.
